control-plane/agents/src/bin/core/volume/specs.rs

```diff
@@ -373,6 +373,43 @@ impl ResourceSpecs {
 
         PaginatedResult::new(self.volumes.paginate(offset, length), last_result)
     }
+
+    pub(crate) fn paginated_snapshots(
+        &self,
+        pagination: &Pagination,
+        vol_id: Option<&VolumeId>,
+    ) -> PaginatedResult<VolumeSnapshot> {
+        let mut last_result = false;
+        if let Some(vol_id) = vol_id {
+            // We need to filter the resource map based on volume id and apply the pagination
+            // parameters on that.
+            let filtered_snapshots = self.volume_snapshots.paginate_selected(
```

Comment:
Here is the usage of new paginate function.  @tiagolobocastro @Abhinandan-Purkait